Change ads campaign creation and editing to use wp-data

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Closes #445 

It also has a temporary fix for #1210 on the client side.

- Add `createAdsCampaign` action and relevant handling to *~/data*.
- Change the campaign creation to use `createAdsCampaign` action.
- Change the campaign editing to use` useAdsCampaigns` hook and `updateAdsCampaign` action.
- Temporarily fix that client-side doesn't correctly show the handling of campaign creation/edit failure.

### Screenshots:

https://user-images.githubusercontent.com/17420811/150722559-08f1e4de-95aa-42a7-b046-92a20baf9f6e.mp4

### Detailed test instructions:

💡  To test the invalid parameters when creating/editing ads campaign, it'd be easier by changing store currency to "Japanese yen (¥)" in the WC general settings.

#### Google Ads account setup

1. Disconnect connection of Google Ads account.
1. Go to Google Ads account setup page.
1. Enter **0.01** in the daily average cost field in step 2.
1. Click "Launch paid campaign" in step 3. It should show a failure notification.
1. Back to step 2 and enter **100** or more in the daily average cost field.
1. Click "Launch paid campaign" in step 3. It should be able to finish the setup.
1. After being redirected to dashboard page, the created campaign should be shown on the **Programs** table.

#### Ads campaign creation

1. Go to the ads campaign creation page.
1. Enter **0.01** in the daily average cost field.
1. Click on "Launch paid campaign" button. It should show a failure notification.
1. enter **100** or more in the daily average cost field.
1. Click on "Launch paid campaign" button. It should be able to create a new campaign.
1. After being redirected to dashboard page, the created campaign should be shown on the **Programs** table.

#### Ads campaign edit

1. Pick an existing campaign to edit.
1. Enter **0.01** in the daily average cost field.
1. Click on "Save changes" button. It should show a failure notification.
1. enter **100** or more in the daily average cost field.
1. Click on "Save changes" button. It should be able to update campaign.
1. After being redirected to dashboard page, that campaign should be shown on the **Programs** table with the updated daily budget.

### Changelog entry

> Fix - Failure handling was not correctly displayed when the ads campaign creation and editing failed.

eason9487 committed:
Add createAdsCampaign action and relevant handling to ~/data

eason9487 committed:
Change the campaign creation to use createAdsCampaign action

eason9487 committed:
Change the campaign edit to use useAdsCampaigns hook and updateAdsCampaign action

eason9487 committed:
Temporarily fix that client-side doesn't correctly show the handling of campaign creation/edit failure.

Partial resolve: https://github.com/woocommerce/google-listings-and-ads/issues/1210

puntope commented on js/src/data/actions.js:
```diff
@@ -765,14 +765,85 @@ export function receiveAdsCampaigns( adsCampaigns ) {
+			 *
+			 * TODO: Remove this code block after the API responds the fails in a more specific way.
+			 */
+			const keys = Object.keys( createdCampaign );
+			if ( keys.length === 1 && keys.pop() === 'message' ) {
```
💅  

These lines together with lines 841-842 could be refactored in a function. In the future, instead of removing the block, we can just adapt that function.

eason9487 replied:
I was leaning toward adjusting them after the #1210 is fixed and not to refactor workarounds based on uncertain future changes for two reasons:
1. They are using different APIs and the APIs are also independent implementations.
2. I assumed the fixing direction of APIs would be more likely to respond with HTTP status code 400 or something like that. And then we just need to delete the whole block without having to check other function references and whether that functions are shared elsewhere.

puntope replied:
> And then we just need to delete the whole block without having to check other function references and whether that functions are shared elsewhere.

If we're going to use the function elsewhere, we shouldn't delete it. We just need to delete/update the line calling the function once the API response is fixed. Besides that, I think is harder to locate the block duplicated over the code base than a named function. 

 > They are using different APIs and the APIs are also independent implementations.

In that case, is more likely to happen that one of them is fixed and then the other implementation is fixed after. Hence makes more sense for me to have a temporary function.  

> I was leaning toward adjusting them after the #1210 is fixed

I'm fine having this refactor in the future since this is not a blocker. But I think would be more correct to have it in a function :)

eason9487 replied:
Sorry, I don't really understand your thoughts.

> If we're going to use the function elsewhere, we shouldn't delete it. We just need to delete/update the line calling the function once the API response is fixed.

This temporary workaround is not intended to reuse elsewhere. And I don't see why the code needs to be deleted is not clear enough. There are code comments and **TODO** flag, and the codes are wrapped by a block.

https://github.com/woocommerce/google-listings-and-ads/blob/92b041a406d016c539c86bb6562e3f04f83dfdd4/js/src/data/actions.js#L787-L798

Would it more clearer for you if I add those code blocks to #1210 once this PR gets merged?

puntope replied:
At the beginning, it was suggested to refactor these blocks in a function i.e `temporaryCheckFunctionResponse(createdCampaign)`

One of the reasons mentioned to don't do so was this:

> 2. I assumed the fixing direction of APIs would be more likely to respond with HTTP status code 400 or something like that. And then we just need to delete the whole block without having to check other function references and whether that functions are shared elsewhere.

Maybe I understood it wrong. But this is what I get for that sentence:

- The code is going to be deleted or modified 
- The code is going to be harder to delete or modify if we refactor it as a function. Because we would need to check function references and whether those functions are shared elsewhere -> This part I disagree... since it would be easier the modification in the form of a function than a duplicated block as is now.

In brief, my thoughts are that right now, the code is duplicated and could be refactored.  I understand the intent of the code and its temporary nature. But I disagree about that temporary nature is a justification to have it duplicated.

eason9487 replied:
Thanks for the explanation.

> * The code is going to be deleted or modified
> * [...] This part I disagree... since it would be easier the modification in the form of a function than a duplicated block as is now.

Based on most of current APIs will respond with status code 400 when parameters are invalid. I assumed most likely the code block would be deleted, not modified. Even if the error is later specified in a different way, how it should be refactored would be clearer at that point. Considering whether to refactor it now seems to be a bit over-engineering to me.

In addition, a small amount of a few lines of duplicated code does not always have a bad smell, sometimes it brings nicer readability to code fragments, here I think the main consideration is how much it burdens the reader's mind, more than whether it is very much aligned with the clean code styles, DRY concepts, or whether it is easy to maintain, since it would most probably be deleted in its entirety.

> In brief, my thoughts are that right now, the code is duplicated and could be refactored. I understand the intent of the code and its temporary nature. But I disagree about that temporary nature is a justification to have it duplicated.

I would like to wrap up this thread with the current implementation and proceed with PR merging. The above are mainly my subjective thoughts, but not a strong opinion. And I know the code is not ideal in some perspectives. I would be happy to adopt your proposed implementation, but I have no idea what's the ideal implementation you thought to refactor it. If you think it's necessary, I would appreciate it if you could open a PR for it. 🙏

puntope commented on js/src/data/test/reducer.test.js:
```diff
@@ -349,6 +349,27 @@ describe( 'reducer', () => {
 			expect( state ).toHaveProperty( path, action.adsCampaigns );
 		} );
 
+		it( 'should push the created ads campaign and return with updated ads campaigns', () => {
+			const createdToInitialState = reducer( prepareState(), {
+				type: TYPES.CREATE_ADS_CAMPAIGN,
+				createdCampaign: { id: 123 },
+			} );
+			const createdToLoadedState = reducer( createdToInitialState, {
+				type: TYPES.CREATE_ADS_CAMPAIGN,
+				createdCampaign: { id: 456 },
+			} );
+
+			createdToInitialState.assertConsistentRef();
+			createdToLoadedState.assertConsistentRef();
+			expect( createdToInitialState ).toHaveProperty( path, [
+				{ id: 123 },
+			] );
+			expect( createdToLoadedState ).toHaveProperty( path, [
+				{ id: 123 },
+				{ id: 456 },
+			] );
+		} );
```
❓ WDYT about adding some test case to check the error handling you added in the PR? (Same with the create Campaign case)

eason9487 replied:
Currently, the errors won't be passed to reducer. Do you mean the test cases for `createAdsCampaign` and `updateAdsCampaign` actions?

Assuming you were meaning actions, I'm not sure about that. Because currently the APIs do not respond with HTTP status 400 or something like that when the parameters are invalid (see #1210). For example, should `apiFetch` mock simulate the problem of current APIs, or should it simulate a non-existent HTTP status 400 response?

Given that neither of these is really certain, I thought we could wait for the #1210 API to be fixed and then test it accordingly to make it robust and less confusing.

puntope replied:
> Currently, the errors won't be passed to reducer. Do you mean the test cases for createAdsCampaign and updateAdsCampaign actions?

What I was looking for was to have a case in which the API call fails and we check at least that it is actually throwing an error.  As far as I understand this was one of the intends of this PR. 

> For example, should apiFetch mock simulate the problem of current APIs

I think the best would be to test always with the current status and then update the code and tests after the final fix.

> Given that neither of these is really certain, I thought we could wait for the #1210 API to be fixed and then test it accordingly to make it robust and less confusing.

I'm unsure when it is expected to have the issue fixed, but if you consider it is appropriate to wait, let's wait until the final fix is done.

puntope commented on js/src/pages/edit-paid-ads-campaign/index.js:
```diff
@@ -9,23 +9,30 @@ import { __, sprintf } from '@wordpress/i18n';
+		data: campaign || null,
+	};
+}
+
```
❓ WDYT about having this hook in a separated file with their own tests?

puntope replied:
💅  Besides that, IMO this function is not giving us a lot. Or I don't see the advantage of it

Actually, I'd just go for... 

```
const { loaded, data: campaigns } = useAdsCampaigns();
const campaign = campaigns?.find( ( el ) => el.id == programId );
```

In the `EditPaidAdsCampaign` component

eason9487 replied:
Adjusted by https://github.com/woocommerce/google-listings-and-ads/pull/1211/commits/92b041a406d016c539c86bb6562e3f04f83dfdd4.

puntope reviewed (commented):
✅  LGTM 

- Tested Google Ads account setup. Works as expected
- Tested Google Ads Add Campaign. Works as expected
- Tested Google Ads Edit Campaign. Works as expected

💅 Some opinionated comments/questions

eason9487 committed:
Internalize `useAdsCampaign` into EditPaidAdsCampaign component.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1211#discussion_r790554282
